File: packages/azalee/src/article-series.ts
/**
 * Portable article-series rules for the Azalée editorial product.
 *
 * The plans intentionally describe values and ordering only. Next actions,
 * API routes, desktop adapters, and background jobs choose their own
 * authenticated transport; no Supabase, Postgres, or framework dependency is
 * allowed here.
 */

export const ARTICLE_SERIES_IDENTIFIER_MAX_LENGTH = 512;
export const ARTICLE_SERIES_ORDER_MIN = 0;
export const ARTICLE_SERIES_ORDER_MAX = 2_147_483_647;

export const ARTICLE_SERIES_SELECT =
	"id, title, slug, description, cover_image_url, author_id";
export const ARTICLE_SERIES_ARTICLE_SELECT =
	"id, title, slug, excerpt, featured_image_url, category, published_at, series_order";

export interface SeriesInfo {
	id: string;
	title: string;
	slug: string;
	description: string | null;
	cover_image_url: string | null;
	author_id: string | null;
}

export interface SeriesArticle {
	id: string;
	title: string;
	slug: string;
	excerpt: string | null;
	featured_image_url: string | null;
	category: string | null;
	published_at: string | null;
	series_order: number | null;
}

export interface SeriesWithArticles {
	series: SeriesInfo;
	articles: SeriesArticle[];
}

export interface SeriesSummary {
	id: string;
	title: string;
	slug: string;
	description: string | null;
	cover_image_url: string | null;
	article_count: number;
}

export interface CreateSeriesInput {
	title: string;
	slug: string;
	description?: string;
	cover_image_url?: string;
}

export interface SeriesBySlugPlan {
	slug: string;
}

export interface SeriesByIdPlan {
	id: string;
}

export interface SeriesArticlesPlan {
	seriesId: string;
	ascending: true;
	orderColumn: "series_order";
}

export interface SeriesArticleCountPlan {
	seriesId: string;
}

export interface CreateSeriesPlan {
	payload: {
		title: string;
		slug: string;
		description: string | null;
		cover_image_url: string | null;
	};
}

export interface AssignArticleToSeriesPlan {
	articleId: string;
	seriesId: string;
	seriesOrder: number;
}

export interface RemoveArticleFromSeriesPlan {
	articleId: string;
	seriesId: null;
	seriesOrder: 0;
}

function normalizeIdentifier(value: unknown, label: string): string {
	if (typeof value !== "string") {
		throw new TypeError(`An article series ${label} must be a string`);
	}

	const identifier = value.trim();
	if (
		!identifier ||
		identifier.length > ARTICLE_SERIES_IDENTIFIER_MAX_LENGTH ||
		identifier.includes("\0")
	) {
		throw new RangeError(`Invalid article series ${label}`);
	}

	return identifier;
}

function normalizeOptionalText(value: unknown, label: string): string | null {
	if (value === undefined || value === null) return null;
	if (typeof value !== "string" || value.includes("\0")) {
		throw new TypeError(`An article series ${label} must be text when provided`);
	}
	return value;
}

/** Normalizes a storage identity without requiring a UUID implementation. */
export function normalizeSeriesId(value: unknown): string {
	return normalizeIdentifier(value, "identifier");
}

/** Normalizes a public series slug before a transport turns it into a query. */
export function normalizeSeriesSlug(value: unknown): string {
	return normalizeIdentifier(value, "slug");
}

/** Normalizes an article identity without prescribing its backing store. */
export function normalizeSeriesArticleId(value: unknown): string {
	return normalizeIdentifier(value, "article identifier");
}

/** Validates the signed 32-bit sequence accepted by the existing SQL schema. */
export function normalizeSeriesOrder(value: unknown): number {
	if (
		typeof value !== "number" ||
		!Number.isInteger(value) ||
		value < ARTICLE_SERIES_ORDER_MIN ||
		value > ARTICLE_SERIES_ORDER_MAX
	) {
		throw new RangeError("Invalid article series order");
	}
	return value;
}

/** Validates and plans the insert payload without coupling it to PostgREST. */
export function planCreateSeries(input: CreateSeriesInput): CreateSeriesPlan {
	return {
		payload: {
			cover_image_url: normalizeOptionalText(input.cover_image_url, "cover image URL"),
			description: normalizeOptionalText(input.description, "description"),
			slug: normalizeSeriesSlug(input.slug),
			title: normalizeIdentifier(input.title, "title"),
		},
	};
}

/** Plans a public series lookup by its stable slug. */
export function planSeriesBySlug(slug: unknown): SeriesBySlugPlan {
	return { slug: normalizeSeriesSlug(slug) };
}

/** Plans an identity lookup used after an article reveals its series ID. */
export function planSeriesById(id: unknown): SeriesByIdPlan {
	return { id: normalizeSeriesId(id) };
}

/** Plans the ordered article collection that backs a series navigator. */
export function planSeriesArticles(seriesId: unknown): SeriesArticlesPlan {
	return {
		ascending: true,
		orderColumn: "series_order",
		seriesId: normalizeSeriesId(seriesId),
	};
}

/** Plans a count request for one series. */
export function planSeriesArticleCount(seriesId: unknown): SeriesArticleCountPlan {
	return { seriesId: normalizeSeriesId(seriesId) };
}

/** Plans a membership update while preserving the storage field semantics. */
export function planAssignArticleToSeries(
	articleId: unknown,
	seriesId: unknown,
	seriesOrder: unknown,
): AssignArticleToSeriesPlan {
	return {
		articleId: normalizeSeriesArticleId(articleId),
		seriesId: normalizeSeriesId(seriesId),
		seriesOrder: normalizeSeriesOrder(seriesOrder),
	};
}

/** Plans removal with the legacy zero order retained for compatibility. */
export function planRemoveArticleFromSeries(articleId: unknown): RemoveArticleFromSeriesPlan {
	return {
		articleId: normalizeSeriesArticleId(articleId),
		seriesId: null,
		seriesOrder: 0,
	};
}

/**
 * Returns a new ordered collection. Equal and missing order values preserve
 * their source order, matching the historical UI and database response.
 */
export function orderSeriesArticles<T extends Pick<SeriesArticle, "series_order">>(
	articles: readonly T[],
): T[] {
	return [...articles].sort(
		(left, right) => (left.series_order ?? 0) - (right.series_order ?? 0),
	);
}

/** Builds the public payload while applying the single shared order rule. */
export function toSeriesWithArticles(
	series: SeriesInfo,
	articles: readonly SeriesArticle[],
): SeriesWithArticles {
	return { articles: orderSeriesArticles(articles), series };
}

/** Maps the ordered collection position for a host-neutral series navigator. */
export function seriesArticleIndex(articles: readonly SeriesArticle[], articleId: unknown): number {
	const normalizedArticleId = normalizeSeriesArticleId(articleId);
	const index = orderSeriesArticles(articles).findIndex((article) => article.id === normalizedArticleId);
	return index >= 0 ? index : 0;
}

/** Combines counted rows with their public summary DTOs without leaking maps. */
export function toSeriesSummaries(
	series: readonly Omit<SeriesSummary, "article_count">[],
	articleCounts: ReadonlyMap<string, number>,
): SeriesSummary[] {
	return series.map((entry) => ({
		...entry,
		article_count: articleCounts.get(entry.id) ?? 0,
	}));
}

File: crates/engine/nie-core/src/azalee/article_series.rs
//! Portable article-series identity and ordering rules.
//!
//! Storage queries and mutations remain in adapters. This layer only validates
//! keys and applies the shared editorial ordering rule.

/// Maximum series and article identifier length.
pub const SERIES_IDENTIFIER_MAX_LENGTH: usize = 512;
/// First valid series order value.
pub const SERIES_ORDER_MIN: i64 = 0;
/// Last valid series order value (historical storage `INT4`).
pub const SERIES_ORDER_MAX: i64 = i32::MAX as i64;

/// Minimal article used by the series navigator.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SeriesArticle {
    /// Article identity.
    pub id: String,
    /// Public title.
    pub title: String,
    /// Public slug.
    pub slug: String,
    /// Optional position in the series.
    pub series_order: Option<i32>,
}

/// Series and articles in editorial order.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SeriesWithArticles {
    /// Public series identity.
    pub series_id: String,
    /// Ordered articles.
    pub articles: Vec<SeriesArticle>,
}

/// Series validation errors.
#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum SeriesError {
    /// A key is empty, too long, or contains a NUL.
    #[error("invalid article series {field}")]
    InvalidIdentifier {
        /// Rejected key category.
        field: &'static str,
    },
    /// A position is outside the historical range.
    #[error("invalid article series order")]
    InvalidOrder,
}

fn normalize_identifier(value: &str, field: &'static str) -> Result<String, SeriesError> {
    let value = value.trim();
    if value.is_empty() || value.len() > SERIES_IDENTIFIER_MAX_LENGTH || value.contains('\0') {
        return Err(SeriesError::InvalidIdentifier { field });
    }
    Ok(value.to_owned())
}

/// Validates a series identity.
pub fn normalize_series_id(value: &str) -> Result<String, SeriesError> {
    normalize_identifier(value, "identifier")
}

/// Validates a public series slug without rewriting it.
pub fn normalize_series_slug(value: &str) -> Result<String, SeriesError> {
    normalize_identifier(value, "slug")
}

/// Validates an article identity belonging to a series.
pub fn normalize_series_article_id(value: &str) -> Result<String, SeriesError> {
    normalize_identifier(value, "article identifier")
}

/// Validates a position compatible with the historical SQL schema.
pub fn normalize_series_order(value: i64) -> Result<i32, SeriesError> {
    if !(SERIES_ORDER_MIN..=SERIES_ORDER_MAX).contains(&value) {
        return Err(SeriesError::InvalidOrder);
    }
    Ok(value as i32)
}

/// Returns a copy sorted by ascending position.
///
/// Missing positions are zero and ties preserve source order.
#[must_use]
pub fn order_series_articles(articles: &[SeriesArticle]) -> Vec<SeriesArticle> {
    let mut ordered = articles.to_vec();
    ordered.sort_by_key(|article| article.series_order.unwrap_or_default());
    ordered
}

/// Builds the public series view using the single ordering rule.
pub fn to_series_with_articles(
    series_id: &str,
    articles: &[SeriesArticle],
) -> Result<SeriesWithArticles, SeriesError> {
    Ok(SeriesWithArticles {
        series_id: normalize_series_id(series_id)?,
        articles: order_series_articles(articles),
    })
}

/// Returns an article position, or zero when the article is absent.
pub fn series_article_index(
    articles: &[SeriesArticle],
    article_id: &str,
) -> Result<usize, SeriesError> {
    let article_id = normalize_series_article_id(article_id)?;
    Ok(order_series_articles(articles)
        .iter()
        .position(|article| article.id == article_id)
        .unwrap_or_default())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn article(id: &str, order: Option<i32>) -> SeriesArticle {
        SeriesArticle {
            id: id.to_owned(),
            title: id.to_owned(),
            slug: id.to_owned(),
            series_order: order,
        }
    }

    #[test]
    fn validates_series_keys_and_int32_order() {
        assert_eq!(normalize_series_id(" s1 ").unwrap(), "s1");
        assert_eq!(
            normalize_series_order(i64::from(i32::MAX)).unwrap(),
            i32::MAX
        );
        assert_eq!(normalize_series_order(-1), Err(SeriesError::InvalidOrder));
        assert_eq!(
            normalize_series_order(i64::from(i32::MAX) + 1),
            Err(SeriesError::InvalidOrder)
        );
        assert!(matches!(
            normalize_series_slug(""),
            Err(SeriesError::InvalidIdentifier { field: "slug" })
        ));
    }

    #[test]
    fn orders_missing_values_first_and_preserves_equal_order() {
        let articles = vec![
            article("third", Some(2)),
            article("missing", None),
            article("first", Some(1)),
            article("zero", Some(0)),
        ];
        let ordered = order_series_articles(&articles);
        assert_eq!(
            ordered
                .iter()
                .map(|article| article.id.as_str())
                .collect::<Vec<_>>(),
            ["missing", "zero", "first", "third"]
        );
        assert_eq!(series_article_index(&articles, "first").unwrap(), 2);
        assert_eq!(series_article_index(&articles, "unknown").unwrap(), 0);
    }

    #[test]
    fn builds_a_validated_series_view() {
        let view = to_series_with_articles("series-1", &[article("a", Some(1))]).unwrap();
        assert_eq!(view.series_id, "series-1");
        assert_eq!(view.articles[0].id, "a");
    }
}
